heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/CacheCore.java

```diff
@@ -0,0 +1,543 @@
+                                    long startTime, long endTime, long bucketId,
+                                    MetricsFilter.MetricAggregationType type) {
+    LOG.info("getAggregatedMetrics " + startTime + " " + endTime);
+    // per request
+    long countAvg = 0;
+
+    MetricsCacheQueryUtils.MetricTimeRangeValue rangeValue =
+        new MetricsCacheQueryUtils.MetricTimeRangeValue();
+    rangeValue.startTime = Long.MAX_VALUE;
+    rangeValue.endTime = 0;
+
+    double result = 0;
+    Long startKey = cacheMetric.floorKey(startTime);
+    for (Long key = startKey != null ? startKey : cacheMetric.firstKey();
+         key != null && key <= endTime;
+         key = cacheMetric.higherKey(key)) {
+      LinkedList<MetricDatapoint> bucket = cacheMetric.get(key).get(bucketId);
+
+      if (bucket != null) {
+        for (MetricDatapoint datapoint : bucket) {
+          if (startTime <= datapoint.timestamp && datapoint.timestamp <= endTime) {
+            switch (type) {
+              case AVG:
+                countAvg++;
+              case SUM:
+                result += Double.parseDouble(datapoint.value);
+                break;
+              case LAST:
+                if (rangeValue.endTime < datapoint.timestamp) {
+                  rangeValue.value = datapoint.value;
+                }
+                break;
+              case UNKNOWN:
+              default:
+                LOG.warning("Unknown metric type " + type);
+                return;
+            }
+            rangeValue.startTime = Math.min(rangeValue.startTime, datapoint.timestamp);
+            rangeValue.endTime = Math.max(rangeValue.endTime, datapoint.timestamp);
+          }
+        } // end bucket
+      }
+
+    } // end tree
+
+    if (type.equals(MetricsFilter.MetricAggregationType.AVG)) {
+      rangeValue.value = String.valueOf(result / countAvg);
+    } else if (type.equals(MetricsFilter.MetricAggregationType.SUM)) {
+      rangeValue.value = String.valueOf(result);
+    }
+    if (rangeValue.value != null) {
+      metricValue.add(rangeValue);
+    }
+    LOG.info(rangeValue.toString());
+  }
+
+  /**
+   * for internal process use
+   */
+  public MetricsCacheQueryUtils.ExceptionResponse getExceptions(
+      MetricsCacheQueryUtils.ExceptionRequest request) {
+    synchronized (CacheCore.class) {
+      MetricsCacheQueryUtils.ExceptionResponse response =
+          new MetricsCacheQueryUtils.ExceptionResponse();
+
+      for (Map.Entry<String, Set<String>> entry : request.componentNameInstanceId.entrySet()) {
+        String componentName = entry.getKey();
+        for (String instanceId : entry.getValue()) {
+          int idx = idxComponentInstance.get(componentName).get(instanceId);
+          response.exceptionDatapointList.addAll(cacheException.get(idx));
+        }
+      }
+
+      return response;
+    }
+  }
+
+  public void purge() {
+    long now = System.currentTimeMillis();
+    LOG.info("purge " + now);
```

Comment:
Does purge happen on a regular schedule? If so, we probably don't want logs to fill with "purge some_timestamp".